### PR TITLE
Adds Tests to the Verkle Tree

### LINKED ISSUE #16: Testing the Verkle Trie Implementation
# Testing
The current **_Verkle Trie_** implementation is capable of all the trie operations but does not support the **Stateless** feature, it is yet to be added.

### Tests
The correctness of the tree root commiment have already been tested w.r.t `go-verkle` 
For testing the on the broader perspective we are supposed to test the follows
- [x] Tree Insertions correctness and edge cases
- [x] Cached commitment checks in Branch Nodes ( Internal Nodes )
- [x] Deletion operations 

For test vectors, it can be generated on our own also few can be taken from the `go-verkle` implementation from `tree_test.go`.

### PR DESCRIPTION
**Pull request by advaita-saha**
Fixes #16 

This PR does a complete testing for the **_Verkle Tree_** implementation without the stateless feature, which is not yet implemented. The following testing have been added and performed
- Tree Insertion Tests
  - Direct Insertion Into the Root
  - Insertion of two leaved and matching with their expected positions in the tree
  - Insertions in the last level
- Commitment Tests
  - Cached Commitment Tests to check if any nodes have stale commitment or mistakenly clears commitment
- Tree Deletion Tests
  - Leaf node deletion
  - Deletion of non-existent node
  - Data pruning, commitment changes after deletion
  - Delete for keys with unequal path, but similar till a level

### CODE REVIEW

**Comment by advaita-saha** [comment]
https://github.com/status-im/nim-eth-verkle/blob/ce67f2766d2f4d36ea353c3f72db023ffa3ca08a/tests/test_tree.nim#L237-L258

This tests is failing, the doAssert is commented out for now, and the result of the deletion is echoed. Details of this issue is mentioned here https://github.com/status-im/nim-eth-verkle/issues/16#issuecomment-1902184431

**PS : Fixed**

File: tests/test_tree.nim [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+    tree.updateAllCommitments()
+
+    var newRoot = tree.commitment
+    var newInternal = ((BranchesNode)tree.branches[4]).commitment
+
+    doAssert oldRoot.serializePoint() != newRoot.serializePoint(), "root has stale commitment"
+    doAssert oldInternal.serializePoint() != newInternal.serializePoint(), "internal node has stale commitment"
+
+    # TODO: make the nil check work
+    # doAssert isNil(BranchesNode(tree.branches[1]).commitment), "internal node has mistakenly cleared cached commitment"
```
Comment (advaita-saha) [inline tests/test_tree.nim:106]: @daniellm @agnxsh 
here when I read the commitment it is actually null, but I can't make the `isNil()` work nor `== nil` is working

It's not a necessity but if you can think of something, suggest the change

**Comment by advaita-saha** [comment]
This PR now also adds fixes to the following

- Fix for deletion erasing wrong key https://github.com/status-im/nim-eth-verkle/issues/16#issuecomment-1902184431
- Fix for fetching wrong key
